dask_jobqueue/core.py

```diff
@@ -219,29 +301,37 @@ def stop_workers(self, workers):
         """ Stop a list of workers"""
         if not workers:
             return
-        workers = list(map(int, workers))
-        jobs = [self.jobs[w] for w in workers]
+        jobs = {_job_id_from_worker_name(w.name) for w in workers}
+        self.stop_jobs(jobs)
+
+    def stop_jobs(self, jobs):
+        """ Stop a list of jobs"""
         self._call([self.cancel_command] + list(jobs))
-        for w in workers:
-            with ignoring(KeyError):
-                del self.jobs[w]
 
     def scale_up(self, n, **kwargs):
         """ Brings total worker count up to ``n`` """
-        return self.start_workers(n - len(self.jobs))
+        active_and_pending = sum([len(j.workers) for j in
+                                  self.running_jobs.values()])
+        active_and_pending += self.worker_processes * len(self.pending_jobs)
+        return self.start_workers(n - active_and_pending)
 
     def scale_down(self, workers):
         ''' Close the workers with the given addresses '''
-        if isinstance(workers, dict):
-            names = {v['name'] for v in workers.values()}
-            job_ids = {name.split('-')[-2] for name in names}
-            self.stop_workers(job_ids)
+        workers = []
```

Comment:
After first tests on a real cluster, I've detected a problem in scale_down. I had to look closely for several minutes before seeing the problem: the `workers` argument is erased by a local variable just below!

Scaling up works fine. Did not try adaptive.

As noted before, this is also a small breaking change with before. I need to get used to it, but I believe this is a good change. We will need to explain the behavior somewhere though. It can be weird to have our asked workers rouned up every time we ask for some more.

Reply:
Good catch. Not sure how this crept in.